includes/discovery/services.inc.php

```diff
@@ -1,35 +1,54 @@
 <?php
+/**
+ * services.inc.php
+ *
+ * Creates the correct handler for the trap and then sends it the trap.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Autodiscovers TCP services on devices in the 'server' device type.
+ * To use set in config.php:
+ * [discover_services] = 'true';
+ *
+ * @package    LibreNMS
+ * @link       http://librenms.org
+ * @copyright  2019 KanREN, Inc.
+ * @author     Heath Barnhart <hbarnhart@kanren.net>
+ */
 
 use LibreNMS\Config;
 
 if (Config::get('discover_services')) {
-    // FIXME: use /etc/services?
-    $known_services = array(
-        22  => 'ssh',
-        25  => 'smtp',
-        53  => 'dns',
-        80  => 'http',
-        110 => 'pop',
-        143 => 'imap',
-    );
-
     // Services
     if ($device['type'] == 'server') {
-        $oids = trim(snmp_walk($device, '.1.3.6.1.2.1.6.13.1.1.0.0.0.0', '-Osqn'));
+        $oidV4 = trim(snmp_walk($device, '.1.3.6.1.2.1.6.20.1.4.1.4.0.0.0.0', '-Osqn'));
```

Comment:
I see this is specifically only checking ones on 0.0.0.0.  Any consideration for checking all and ignoring services on 127.x.x.x?

Reply:
Actually, you could just walk TCP-MIB::tcpListenerProcess and get both ipv4 and ipv6.  But you would have to discard localhost listening ones still.
You could also walk UDP-MIB::udpEndpointProcess and add udp services.

discover_service() should be where the checks are if it is a valid check_xxxx file...

Reply:
I should have replied her.  Getting the hang of GitHub.  You might only want to auto-discover services listening on the any addresses rather then all IPs and discarding localhost.  You can manually add a service if its listening on a special IP and you want to monitor it.  There are times that services might not be listening on the any address because they are bound to some Out of Band management network or some special interface like for docker, but that would not be reachable by the LibreNMS monitoring host, or at least not on the regular IP Address a host is monitored with.  

I think there would be more cases of having to turn off services that shouldn't be monitored then adding services that you want monitored, if you grabbed IPs other the the Any IPs. (minus localhost)

Reply:
I don't feel strongly one way or the other, just trying to provide another point of view.